internal/kubeapi/fake/fakeclients.go

```diff
@@ -0,0 +1,123 @@
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+import (
+	"errors"
+	"io/ioutil"
+	"os"
+
+	"github.com/crunchydata/postgres-operator/internal/config"
+	fakecrunchy "github.com/crunchydata/postgres-operator/pkg/generated/clientset/versioned/fake"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	fakekube "k8s.io/client-go/kubernetes/fake"
+)
+
+const (
+	// defaultPGOInstallationName is the default installation name for a fake PGO client
+	defaultPGOInstallationName = "test"
+	// pgoNamespace is the default operator namespace for a fake PGO client
+	defaultPGONamespace = "pgo"
+	// defaultTargetNamespaces are the default target namespaces for a fake PGO client
+	defaultTargetNamespaces = "pgouser1,pgouser2"
+)
+
+var (
+	// pgoRoot represents the root of the PostgreSQL Operator project repository
+	pgoRoot = os.Getenv("PGOROOT")
+	// templatePath defines the default location for the PostgreSQL Operator templates relative to
+	// env var PGOROORT
+	templatePath = pgoRoot + "/installers/ansible/roles/pgo-operator/files/pgo-configs/"
+	// pgoYAMLPath defines the default location for the default pgo.yaml configuration file
+	// relative to env var PGOROORT
+	pgoYAMLPath = pgoRoot + "/conf/postgres-operator/pgo.yaml"
+)
+
+// NewFakePGOClient creates a fake PostgreSQL Operator client.  Specifically, it creates
+// a fake client containing a 'pgo-config' ConfigMap as needed to initialize the Operator
+// (i.e. call the 'operator' packages 'Initialize()' function).  This allows for the proper
+// initialization of the Operator in various unit tests where the various resources loaded
+// during intialization (e.g. templates, config and/or global variables) are required.
+func NewFakePGOClient() (kubernetes.Interface, error) {
+
+	if pgoRoot == "" {
+		return nil, errors.New("Environment variable PGOROOT must be set to the root directory " +
+			"of the PostgreSQL Operator project repository in order to create a fake client")
+	}
+
+	os.Setenv("CRUNCHY_DEBUG", "false")
+	os.Setenv("NAMESPACE", defaultTargetNamespaces)
+	os.Setenv("PGO_INSTALLATION_NAME", defaultPGOInstallationName)
+	os.Setenv("PGO_OPERATOR_NAMESPACE", defaultPGONamespace)
+
+	// create a fake 'pgo-config' ConfigMap containing the operator templates and pgo.yaml
+	pgoConfig, err := createMockPGOConfigMap(defaultPGONamespace)
+	if err != nil {
+		return nil, err
+	}
+
+	// now create and return a fake client containing the ConfigMap
+	return Clientset{
+		Clientset:    fakekube.NewSimpleClientset(pgoConfig),
+		PGOClientset: fakecrunchy.NewSimpleClientset(),
+	}, nil
```

Comment:
I'm glad to see this type works at all; I think this is the first instance of it. If you find anything about it that can improve, please do.

Reply:
Will do!  Seems to be working well so far :)